runtime/lua/vim/lsp/util.lua

```diff
@@ -474,7 +474,20 @@ function M.apply_text_edits(text_edits, bufnr, offset_encoding)
       end
       has_eol_text_edit = true
     end
-    api.nvim_buf_set_text(bufnr, e.start_row, e.start_col, e.end_row, e.end_col, e.text)
+
+    local last_line = get_line(bufnr, e.end_row)
+
+    -- If the replacement is over full lines it is safer to use api.nvim_buf_set_lines
```

Comment:
Why is it safer?

Reply:
Because it handles the case of `e.end_col` being out of bounds.

Reply:
To add some context this is from the nvim documentation:
```
                                                         *nvim_buf_set_text()*
nvim_buf_set_text({buffer}, {start_row}, {start_col}, {end_row}, {end_col},
                  {replacement})
    Sets (replaces) a range in the buffer

    This is recommended over |nvim_buf_set_lines()| when only modifying parts
    of a line, as extmarks will be preserved on non-modified parts of the
    touched lines.

    Indexing is zero-based. Row indices are end-inclusive, and column indices
    are end-exclusive.

    To insert text at a given `(row, column)` location, use `start_row =
    end_row = row` and `start_col = end_col = col`. To delete the text in a
    range, use `replacement = {}`.

    Prefer |nvim_buf_set_lines()| if you are only adding or deleting entire
    lines.
```
So the set_text is recommended when **only** modifying parts of a line and the set_lines is recommended when replacing whole lines.